// @iobroker/db-objects-redis/build/cjs/lib/objects/constants.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  ACCESS_CREATE: () => ACCESS_CREATE,
  ACCESS_DELETE: () => ACCESS_DELETE,
  ACCESS_EVERY_ALL: () => ACCESS_EVERY_ALL,
  ACCESS_EVERY_EXEC: () => ACCESS_EVERY_EXEC,
  ACCESS_EVERY_READ: () => ACCESS_EVERY_READ,
  ACCESS_EVERY_RW: () => ACCESS_EVERY_RW,
  ACCESS_EVERY_WRITE: () => ACCESS_EVERY_WRITE,
  ACCESS_GROUP_ALL: () => ACCESS_GROUP_ALL,
  ACCESS_GROUP_EXEC: () => ACCESS_GROUP_EXEC,
  ACCESS_GROUP_READ: () => ACCESS_GROUP_READ,
  ACCESS_GROUP_RW: () => ACCESS_GROUP_RW,
  ACCESS_GROUP_WRITE: () => ACCESS_GROUP_WRITE,
  ACCESS_LIST: () => ACCESS_LIST,
  ACCESS_READ: () => ACCESS_READ,
  ACCESS_USER_ALL: () => ACCESS_USER_ALL,
  ACCESS_USER_EXEC: () => ACCESS_USER_EXEC,
  ACCESS_USER_READ: () => ACCESS_USER_READ,
  ACCESS_USER_RW: () => ACCESS_USER_RW,
  ACCESS_USER_WRITE: () => ACCESS_USER_WRITE,
  ACCESS_WRITE: () => ACCESS_WRITE,
  ERRORS: () => ERRORS,
  GROUP_STARTS_WITH: () => GROUP_STARTS_WITH,
  REG_CHECK_ID: () => REG_CHECK_ID,
  SYSTEM_ADMIN_GROUP: () => SYSTEM_ADMIN_GROUP,
  SYSTEM_ADMIN_USER: () => SYSTEM_ADMIN_USER,
  USER_STARTS_WITH: () => USER_STARTS_WITH
});
module.exports = __toCommonJS(constants_exports);
var ERRORS;
(function(ERRORS2) {
  ERRORS2["ERROR_PERMISSION"] = "permissionError";
  ERRORS2["ERROR_NOT_FOUND"] = "Not exists";
  ERRORS2["ERROR_DB_CLOSED"] = "DB closed";
})(ERRORS || (ERRORS = {}));
const SYSTEM_ADMIN_USER = "system.user.admin";
const SYSTEM_ADMIN_GROUP = "system.group.administrator";
const ACCESS_EVERY_EXEC = 1;
const ACCESS_EVERY_WRITE = 2;
const ACCESS_EVERY_READ = 4;
const ACCESS_EVERY_RW = ACCESS_EVERY_WRITE | ACCESS_EVERY_READ;
const ACCESS_EVERY_ALL = ACCESS_EVERY_WRITE | ACCESS_EVERY_READ | ACCESS_EVERY_EXEC;
const ACCESS_GROUP_EXEC = 16;
const ACCESS_GROUP_WRITE = 32;
const ACCESS_GROUP_READ = 64;
const ACCESS_GROUP_RW = ACCESS_GROUP_WRITE | ACCESS_GROUP_READ;
const ACCESS_GROUP_ALL = ACCESS_GROUP_WRITE | ACCESS_GROUP_READ | ACCESS_GROUP_EXEC;
const ACCESS_USER_EXEC = 256;
const ACCESS_USER_WRITE = 512;
const ACCESS_USER_READ = 1024;
const ACCESS_USER_RW = ACCESS_USER_WRITE | ACCESS_USER_READ;
const ACCESS_USER_ALL = ACCESS_USER_WRITE | ACCESS_USER_READ | ACCESS_USER_EXEC;
const ACCESS_WRITE = 2;
const ACCESS_READ = 4;
const ACCESS_LIST = "list";
const ACCESS_DELETE = "delete";
const ACCESS_CREATE = "create";
const REG_CHECK_ID = /[*?[\]]|\$%\$/;
const USER_STARTS_WITH = "system.user.";
const GROUP_STARTS_WITH = "system.group.";
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ACCESS_CREATE,
  ACCESS_DELETE,
  ACCESS_EVERY_ALL,
  ACCESS_EVERY_EXEC,
  ACCESS_EVERY_READ,
  ACCESS_EVERY_RW,
  ACCESS_EVERY_WRITE,
  ACCESS_GROUP_ALL,
  ACCESS_GROUP_EXEC,
  ACCESS_GROUP_READ,
  ACCESS_GROUP_RW,
  ACCESS_GROUP_WRITE,
  ACCESS_LIST,
  ACCESS_READ,
  ACCESS_USER_ALL,
  ACCESS_USER_EXEC,
  ACCESS_USER_READ,
  ACCESS_USER_RW,
  ACCESS_USER_WRITE,
  ACCESS_WRITE,
  ERRORS,
  GROUP_STARTS_WITH,
  REG_CHECK_ID,
  SYSTEM_ADMIN_GROUP,
  SYSTEM_ADMIN_USER,
  USER_STARTS_WITH
});
