// @iobroker/db-states-jsonl/build/cjs/lib/states/statesInMemServerClass.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var statesInMemServerClass_exports = {};
__export(statesInMemServerClass_exports, {
  StatesInMemoryServerClass: () => StatesInMemoryServerClass
});
module.exports = __toCommonJS(statesInMemServerClass_exports);
var import_db_states_redis = require("@iobroker/db-states-redis");
var import_statesInMemServerRedis = require("./statesInMemServerRedis.js");
class StatesInMemoryServerClass extends import_db_states_redis.Client {
  constructor(settings) {
    settings.autoConnect = false;
    super(settings);
    const serverSettings = {
      namespace: settings.namespace ? `${settings.namespace}-Server` : "Server",
      connection: settings.connection,
      logger: settings.logger,
      hostname: settings.hostname,
      connected: () => {
        this.connectDb();
      }
    };
    this.statesServer = new import_statesInMemServerRedis.StatesInMemoryServer(serverSettings);
  }
  async destroy() {
    await super.destroy();
    await this.statesServer.destroy();
  }
  getStatus() {
    return this.statesServer.getStatus();
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  StatesInMemoryServerClass
});
