// @iobroker/js-controller-common-db/build/esm/lib/common/exitCodes.js
export var EXIT_CODES;
(function (EXIT_CODES) {
    EXIT_CODES[EXIT_CODES["NO_ERROR"] = 0] = "NO_ERROR";
    EXIT_CODES[EXIT_CODES["JS_CONTROLLER_STOPPED"] = 1] = "JS_CONTROLLER_STOPPED";
    EXIT_CODES[EXIT_CODES["INVALID_ADAPTER_CONFIG"] = 2] = "INVALID_ADAPTER_CONFIG";
    EXIT_CODES[EXIT_CODES["NO_ADAPTER_CONFIG_FOUND"] = 3] = "NO_ADAPTER_CONFIG_FOUND";
    EXIT_CODES[EXIT_CODES["INVALID_CONFIG_OBJECT"] = 4] = "INVALID_CONFIG_OBJECT";
    EXIT_CODES[EXIT_CODES["INVALID_ADAPTER_ID"] = 5] = "INVALID_ADAPTER_ID";
    EXIT_CODES[EXIT_CODES["UNCAUGHT_EXCEPTION"] = 6] = "UNCAUGHT_EXCEPTION";
    EXIT_CODES[EXIT_CODES["ADAPTER_ALREADY_RUNNING"] = 7] = "ADAPTER_ALREADY_RUNNING";
    EXIT_CODES[EXIT_CODES["INSTANCE_IS_DISABLED"] = 8] = "INSTANCE_IS_DISABLED";
    EXIT_CODES[EXIT_CODES["CANNOT_GZIP_DIRECTORY"] = 9] = "CANNOT_GZIP_DIRECTORY";
    EXIT_CODES[EXIT_CODES["CANNOT_FIND_ADAPTER_DIR"] = 10] = "CANNOT_FIND_ADAPTER_DIR";
    EXIT_CODES[EXIT_CODES["ADAPTER_REQUESTED_TERMINATION"] = 11] = "ADAPTER_REQUESTED_TERMINATION";
    EXIT_CODES[EXIT_CODES["UNKNOWN_PACKET_NAME"] = 12] = "UNKNOWN_PACKET_NAME";
    EXIT_CODES[EXIT_CODES["ADAPTER_REQUESTED_REBUILD"] = 13] = "ADAPTER_REQUESTED_REBUILD";
    EXIT_CODES[EXIT_CODES["CANNOT_CREATE_USER_OR_GROUP"] = 14] = "CANNOT_CREATE_USER_OR_GROUP";
    EXIT_CODES[EXIT_CODES["UNKNOWN_ERROR"] = 15] = "UNKNOWN_ERROR";
    EXIT_CODES[EXIT_CODES["NON_EXISTING_HOST"] = 16] = "NON_EXISTING_HOST";
    EXIT_CODES[EXIT_CODES["CANNOT_READ_INSTANCES"] = 18] = "CANNOT_READ_INSTANCES";
    EXIT_CODES[EXIT_CODES["NO_MULTIPLE_INSTANCES_ALLOWED"] = 19] = "NO_MULTIPLE_INSTANCES_ALLOWED";
    EXIT_CODES[EXIT_CODES["CANNOT_GET_HOST_INFO"] = 20] = "CANNOT_GET_HOST_INFO";
    EXIT_CODES[EXIT_CODES["NO_MULTIPLE_INSTANCES_ALLOWED_ON_HOST"] = 21] = "NO_MULTIPLE_INSTANCES_ALLOWED_ON_HOST";
    EXIT_CODES[EXIT_CODES["NO_CONNECTION_TO_OBJ_DB"] = 22] = "NO_CONNECTION_TO_OBJ_DB";
    EXIT_CODES[EXIT_CODES["NO_CONNECTION_TO_STATES_DB"] = 23] = "NO_CONNECTION_TO_STATES_DB";
    EXIT_CODES[EXIT_CODES["INSTANCE_ALREADY_EXISTS"] = 24] = "INSTANCE_ALREADY_EXISTS";
    EXIT_CODES[EXIT_CODES["CANNOT_INSTALL_NPM_PACKET"] = 25] = "CANNOT_INSTALL_NPM_PACKET";
    EXIT_CODES[EXIT_CODES["CANNOT_EXTRACT_FROM_ZIP"] = 26] = "CANNOT_EXTRACT_FROM_ZIP";
    EXIT_CODES[EXIT_CODES["INVALID_IO_PACKAGE_JSON"] = 27] = "INVALID_IO_PACKAGE_JSON";
    EXIT_CODES[EXIT_CODES["CANNOT_COPY_DIR"] = 28] = "CANNOT_COPY_DIR";
    EXIT_CODES[EXIT_CODES["MISSING_ADAPTER_FILES"] = 29] = "MISSING_ADAPTER_FILES";
    EXIT_CODES[EXIT_CODES["INVALID_NPM_VERSION"] = 30] = "INVALID_NPM_VERSION";
    EXIT_CODES[EXIT_CODES["INVALID_NODE_VERSION"] = 31] = "INVALID_NODE_VERSION";
    EXIT_CODES[EXIT_CODES["INVALID_DEPENDENCY_VERSION"] = 32] = "INVALID_DEPENDENCY_VERSION";
    EXIT_CODES[EXIT_CODES["INVALID_ARGUMENTS"] = 33] = "INVALID_ARGUMENTS";
    EXIT_CODES[EXIT_CODES["INVALID_PASSWORD"] = 34] = "INVALID_PASSWORD";
    EXIT_CODES[EXIT_CODES["INVALID_OS"] = 35] = "INVALID_OS";
    EXIT_CODES[EXIT_CODES["CANNOT_SYNC_FILES"] = 36] = "CANNOT_SYNC_FILES";
    EXIT_CODES[EXIT_CODES["CANNOT_ENABLE_MULTIHOST"] = 37] = "CANNOT_ENABLE_MULTIHOST";
    EXIT_CODES[EXIT_CODES["MISSING_CONFIG_JSON"] = 40] = "MISSING_CONFIG_JSON";
    EXIT_CODES[EXIT_CODES["CANNOT_DELETE_NON_DELETABLE"] = 41] = "CANNOT_DELETE_NON_DELETABLE";
    EXIT_CODES[EXIT_CODES["CANNOT_UPLOAD_DATA"] = 42] = "CANNOT_UPLOAD_DATA";
    EXIT_CODES[EXIT_CODES["CANNOT_SET_OBJECT"] = 43] = "CANNOT_SET_OBJECT";
    EXIT_CODES[EXIT_CODES["CANNOT_UPDATE_VENDOR"] = 44] = "CANNOT_UPDATE_VENDOR";
    EXIT_CODES[EXIT_CODES["INVALID_REPO"] = 45] = "INVALID_REPO";
    EXIT_CODES[EXIT_CODES["CANNOT_UPDATE_LICENSE"] = 46] = "CANNOT_UPDATE_LICENSE";
    EXIT_CODES[EXIT_CODES["CANNOT_DELETE_DEPENDENCY"] = 47] = "CANNOT_DELETE_DEPENDENCY";
    EXIT_CODES[EXIT_CODES["CANNOT_GET_STATES"] = 50] = "CANNOT_GET_STATES";
    EXIT_CODES[EXIT_CODES["ADAPTER_ALREADY_INSTALLED"] = 51] = "ADAPTER_ALREADY_INSTALLED";
    EXIT_CODES[EXIT_CODES["CANNOT_RESTORE_BACKUP"] = 52] = "CANNOT_RESTORE_BACKUP";
    EXIT_CODES[EXIT_CODES["ADAPTER_NOT_FOUND"] = 53] = "ADAPTER_NOT_FOUND";
    EXIT_CODES[EXIT_CODES["MIGRATION_ERROR"] = 78] = "MIGRATION_ERROR";
    EXIT_CODES[EXIT_CODES["CONTROLLER_RUNNING"] = 99] = "CONTROLLER_RUNNING";
    EXIT_CODES[EXIT_CODES["CONTROLLER_NOT_RUNNING"] = 100] = "CONTROLLER_NOT_RUNNING";
    EXIT_CODES[EXIT_CODES["CANNOT_GET_UUID"] = 101] = "CANNOT_GET_UUID";
    EXIT_CODES[EXIT_CODES["CANNOT_GET_REPO_LIST"] = 102] = "CANNOT_GET_REPO_LIST";
    EXIT_CODES[EXIT_CODES["START_IMMEDIATELY_AFTER_STOP"] = 156] = "START_IMMEDIATELY_AFTER_STOP";
    EXIT_CODES[EXIT_CODES["FILE_NOT_FOUND"] = 404] = "FILE_NOT_FOUND";
})(EXIT_CODES || (EXIT_CODES = {}));
