// @datalust/winston-seq/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SeqTransport = void 0;
const seq = __importStar(require("seq-logging"));
const winston_transport_1 = __importDefault(require("winston-transport"));
class SeqTransport extends winston_transport_1.default {
    constructor(opts) {
        super(opts);
        this.logger = new seq.Logger(opts);
        setImmediate(() => this.emit('opened'));
    }
    log(info, next) {
        setImmediate(() => this.emit('logged', info));
        const { level, message, exception, stack } = info, props = __rest(info, ["level", "message", "exception", "stack"]);
        this.logger.emit({
            timestamp: props.timestamp || new Date(),
            level: level,
            messageTemplate: message,
            properties: props,
            exception: exception || stack
        });
        next();
    }
    close() {
        this.logger.close();
        setImmediate(() => this.emit('closed'));
    }
    flush() {
        return this.logger.flush();
    }
}
exports.SeqTransport = SeqTransport;
if (!String.prototype.startsWith) {
    Object.defineProperty(String.prototype, 'startsWith', {
        value: function (search, rawPos) {
            const pos = rawPos > 0 ? rawPos | 0 : 0;
            return this.substring(pos, pos + search.length) === search;
        },
    });
}
