// @iobroker/js-controller-common-db/build/cjs/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var esm_exports = {};
__export(esm_exports, {
  EXIT_CODES: () => import_exitCodes.EXIT_CODES,
  constants: () => constants,
  defaultRedisInterview: () => import_interview.defaultRedisInterview,
  logger: () => import_logger.logger,
  password: () => import_password.password,
  session: () => import_session.createAdapterStore,
  tools: () => tools
});
module.exports = __toCommonJS(esm_exports);
var tools = __toESM(require("./lib/common/tools.js"), 1);
var import_exitCodes = require("./lib/common/exitCodes.js");
var import_password = require("./lib/common/password.js");
var import_logger = require("./lib/common/logger.js");
var import_interview = require("./lib/common/interview.js");
var import_session = require("./lib/common/session.js");
var constants = __toESM(require("./lib/common/constants.js"), 1);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  EXIT_CODES,
  constants,
  defaultRedisInterview,
  logger,
  password,
  session,
  tools
});
