// @iobroker/js-controller-common-db/build/cjs/lib/common/constants.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  DEFAULT_OBJECTS_WARN_LIMIT: () => DEFAULT_OBJECTS_WARN_LIMIT,
  HIGHEST_UNICODE_SYMBOL: () => HIGHEST_UNICODE_SYMBOL,
  SYSTEM_ADAPTER_PREFIX: () => SYSTEM_ADAPTER_PREFIX,
  SYSTEM_CONFIG_ID: () => SYSTEM_CONFIG_ID,
  SYSTEM_HOST_PREFIX: () => SYSTEM_HOST_PREFIX,
  SYSTEM_REPOSITORIES_ID: () => SYSTEM_REPOSITORIES_ID
});
module.exports = __toCommonJS(constants_exports);
const SYSTEM_ADAPTER_PREFIX = "system.adapter.";
const SYSTEM_HOST_PREFIX = "system.host.";
const SYSTEM_REPOSITORIES_ID = "system.repositories";
const SYSTEM_CONFIG_ID = "system.config";
const HIGHEST_UNICODE_SYMBOL = "\u9999";
const DEFAULT_OBJECTS_WARN_LIMIT = 5e3;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  DEFAULT_OBJECTS_WARN_LIMIT,
  HIGHEST_UNICODE_SYMBOL,
  SYSTEM_ADAPTER_PREFIX,
  SYSTEM_CONFIG_ID,
  SYSTEM_HOST_PREFIX,
  SYSTEM_REPOSITORIES_ID
});
