// @iobroker/js-controller-common-db/build/cjs/lib/common/exitCodes.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var exitCodes_exports = {};
__export(exitCodes_exports, {
  EXIT_CODES: () => EXIT_CODES
});
module.exports = __toCommonJS(exitCodes_exports);
var EXIT_CODES;
(function(EXIT_CODES2) {
  EXIT_CODES2[EXIT_CODES2["NO_ERROR"] = 0] = "NO_ERROR";
  EXIT_CODES2[EXIT_CODES2["JS_CONTROLLER_STOPPED"] = 1] = "JS_CONTROLLER_STOPPED";
  EXIT_CODES2[EXIT_CODES2["INVALID_ADAPTER_CONFIG"] = 2] = "INVALID_ADAPTER_CONFIG";
  EXIT_CODES2[EXIT_CODES2["NO_ADAPTER_CONFIG_FOUND"] = 3] = "NO_ADAPTER_CONFIG_FOUND";
  EXIT_CODES2[EXIT_CODES2["INVALID_CONFIG_OBJECT"] = 4] = "INVALID_CONFIG_OBJECT";
  EXIT_CODES2[EXIT_CODES2["INVALID_ADAPTER_ID"] = 5] = "INVALID_ADAPTER_ID";
  EXIT_CODES2[EXIT_CODES2["UNCAUGHT_EXCEPTION"] = 6] = "UNCAUGHT_EXCEPTION";
  EXIT_CODES2[EXIT_CODES2["ADAPTER_ALREADY_RUNNING"] = 7] = "ADAPTER_ALREADY_RUNNING";
  EXIT_CODES2[EXIT_CODES2["INSTANCE_IS_DISABLED"] = 8] = "INSTANCE_IS_DISABLED";
  EXIT_CODES2[EXIT_CODES2["CANNOT_GZIP_DIRECTORY"] = 9] = "CANNOT_GZIP_DIRECTORY";
  EXIT_CODES2[EXIT_CODES2["CANNOT_FIND_ADAPTER_DIR"] = 10] = "CANNOT_FIND_ADAPTER_DIR";
  EXIT_CODES2[EXIT_CODES2["ADAPTER_REQUESTED_TERMINATION"] = 11] = "ADAPTER_REQUESTED_TERMINATION";
  EXIT_CODES2[EXIT_CODES2["UNKNOWN_PACKET_NAME"] = 12] = "UNKNOWN_PACKET_NAME";
  EXIT_CODES2[EXIT_CODES2["ADAPTER_REQUESTED_REBUILD"] = 13] = "ADAPTER_REQUESTED_REBUILD";
  EXIT_CODES2[EXIT_CODES2["CANNOT_CREATE_USER_OR_GROUP"] = 14] = "CANNOT_CREATE_USER_OR_GROUP";
  EXIT_CODES2[EXIT_CODES2["UNKNOWN_ERROR"] = 15] = "UNKNOWN_ERROR";
  EXIT_CODES2[EXIT_CODES2["NON_EXISTING_HOST"] = 16] = "NON_EXISTING_HOST";
  EXIT_CODES2[EXIT_CODES2["CANNOT_READ_INSTANCES"] = 18] = "CANNOT_READ_INSTANCES";
  EXIT_CODES2[EXIT_CODES2["NO_MULTIPLE_INSTANCES_ALLOWED"] = 19] = "NO_MULTIPLE_INSTANCES_ALLOWED";
  EXIT_CODES2[EXIT_CODES2["CANNOT_GET_HOST_INFO"] = 20] = "CANNOT_GET_HOST_INFO";
  EXIT_CODES2[EXIT_CODES2["NO_MULTIPLE_INSTANCES_ALLOWED_ON_HOST"] = 21] = "NO_MULTIPLE_INSTANCES_ALLOWED_ON_HOST";
  EXIT_CODES2[EXIT_CODES2["NO_CONNECTION_TO_OBJ_DB"] = 22] = "NO_CONNECTION_TO_OBJ_DB";
  EXIT_CODES2[EXIT_CODES2["NO_CONNECTION_TO_STATES_DB"] = 23] = "NO_CONNECTION_TO_STATES_DB";
  EXIT_CODES2[EXIT_CODES2["INSTANCE_ALREADY_EXISTS"] = 24] = "INSTANCE_ALREADY_EXISTS";
  EXIT_CODES2[EXIT_CODES2["CANNOT_INSTALL_NPM_PACKET"] = 25] = "CANNOT_INSTALL_NPM_PACKET";
  EXIT_CODES2[EXIT_CODES2["CANNOT_EXTRACT_FROM_ZIP"] = 26] = "CANNOT_EXTRACT_FROM_ZIP";
  EXIT_CODES2[EXIT_CODES2["INVALID_IO_PACKAGE_JSON"] = 27] = "INVALID_IO_PACKAGE_JSON";
  EXIT_CODES2[EXIT_CODES2["CANNOT_COPY_DIR"] = 28] = "CANNOT_COPY_DIR";
  EXIT_CODES2[EXIT_CODES2["MISSING_ADAPTER_FILES"] = 29] = "MISSING_ADAPTER_FILES";
  EXIT_CODES2[EXIT_CODES2["INVALID_NPM_VERSION"] = 30] = "INVALID_NPM_VERSION";
  EXIT_CODES2[EXIT_CODES2["INVALID_NODE_VERSION"] = 31] = "INVALID_NODE_VERSION";
  EXIT_CODES2[EXIT_CODES2["INVALID_DEPENDENCY_VERSION"] = 32] = "INVALID_DEPENDENCY_VERSION";
  EXIT_CODES2[EXIT_CODES2["INVALID_ARGUMENTS"] = 33] = "INVALID_ARGUMENTS";
  EXIT_CODES2[EXIT_CODES2["INVALID_PASSWORD"] = 34] = "INVALID_PASSWORD";
  EXIT_CODES2[EXIT_CODES2["INVALID_OS"] = 35] = "INVALID_OS";
  EXIT_CODES2[EXIT_CODES2["CANNOT_SYNC_FILES"] = 36] = "CANNOT_SYNC_FILES";
  EXIT_CODES2[EXIT_CODES2["CANNOT_ENABLE_MULTIHOST"] = 37] = "CANNOT_ENABLE_MULTIHOST";
  EXIT_CODES2[EXIT_CODES2["MISSING_CONFIG_JSON"] = 40] = "MISSING_CONFIG_JSON";
  EXIT_CODES2[EXIT_CODES2["CANNOT_DELETE_NON_DELETABLE"] = 41] = "CANNOT_DELETE_NON_DELETABLE";
  EXIT_CODES2[EXIT_CODES2["CANNOT_UPLOAD_DATA"] = 42] = "CANNOT_UPLOAD_DATA";
  EXIT_CODES2[EXIT_CODES2["CANNOT_SET_OBJECT"] = 43] = "CANNOT_SET_OBJECT";
  EXIT_CODES2[EXIT_CODES2["CANNOT_UPDATE_VENDOR"] = 44] = "CANNOT_UPDATE_VENDOR";
  EXIT_CODES2[EXIT_CODES2["INVALID_REPO"] = 45] = "INVALID_REPO";
  EXIT_CODES2[EXIT_CODES2["CANNOT_UPDATE_LICENSE"] = 46] = "CANNOT_UPDATE_LICENSE";
  EXIT_CODES2[EXIT_CODES2["CANNOT_DELETE_DEPENDENCY"] = 47] = "CANNOT_DELETE_DEPENDENCY";
  EXIT_CODES2[EXIT_CODES2["CANNOT_GET_STATES"] = 50] = "CANNOT_GET_STATES";
  EXIT_CODES2[EXIT_CODES2["ADAPTER_ALREADY_INSTALLED"] = 51] = "ADAPTER_ALREADY_INSTALLED";
  EXIT_CODES2[EXIT_CODES2["CANNOT_RESTORE_BACKUP"] = 52] = "CANNOT_RESTORE_BACKUP";
  EXIT_CODES2[EXIT_CODES2["ADAPTER_NOT_FOUND"] = 53] = "ADAPTER_NOT_FOUND";
  EXIT_CODES2[EXIT_CODES2["MIGRATION_ERROR"] = 78] = "MIGRATION_ERROR";
  EXIT_CODES2[EXIT_CODES2["CONTROLLER_RUNNING"] = 99] = "CONTROLLER_RUNNING";
  EXIT_CODES2[EXIT_CODES2["CONTROLLER_NOT_RUNNING"] = 100] = "CONTROLLER_NOT_RUNNING";
  EXIT_CODES2[EXIT_CODES2["CANNOT_GET_UUID"] = 101] = "CANNOT_GET_UUID";
  EXIT_CODES2[EXIT_CODES2["CANNOT_GET_REPO_LIST"] = 102] = "CANNOT_GET_REPO_LIST";
  EXIT_CODES2[EXIT_CODES2["START_IMMEDIATELY_AFTER_STOP"] = 156] = "START_IMMEDIATELY_AFTER_STOP";
  EXIT_CODES2[EXIT_CODES2["FILE_NOT_FOUND"] = 404] = "FILE_NOT_FOUND";
})(EXIT_CODES || (EXIT_CODES = {}));
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  EXIT_CODES
});
