// @iobroker/socket-client/build/cjs/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var index_exports = {};
__export(index_exports, {
  AdminConnection: () => import_AdminConnection.AdminConnection,
  Connection: () => import_Connection.Connection,
  ERRORS: () => import_Connection.ERRORS,
  NOT_CONNECTED: () => import_Connection.NOT_CONNECTED,
  PERMISSION_ERROR: () => import_Connection.PERMISSION_ERROR,
  PROGRESS: () => import_Connection.PROGRESS
});
module.exports = __toCommonJS(index_exports);
var import_AdminConnection = require("./AdminConnection.js");
var import_Connection = require("./Connection.js");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AdminConnection,
  Connection,
  ERRORS,
  NOT_CONNECTED,
  PERMISSION_ERROR,
  PROGRESS
});
