// @iobroker/type-detector/build/index.js
"use strict";
/**
 * Copyright 2018-2025 bluefox <dogafox@gmail.com>
 *
 * The MIT License (MIT)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StateType = exports.Types = exports.roleOrEnumGate = exports.roleOrEnumDoor = exports.roleOrEnumWindow = exports.roleOrEnumBlind = exports.roleOrEnumLight = void 0;
const ChannelDetector_1 = require("./ChannelDetector");
var roleEnumUtils_1 = require("./roleEnumUtils");
Object.defineProperty(exports, "roleOrEnumLight", { enumerable: true, get: function () { return roleEnumUtils_1.roleOrEnumLight; } });
Object.defineProperty(exports, "roleOrEnumBlind", { enumerable: true, get: function () { return roleEnumUtils_1.roleOrEnumBlind; } });
Object.defineProperty(exports, "roleOrEnumWindow", { enumerable: true, get: function () { return roleEnumUtils_1.roleOrEnumWindow; } });
Object.defineProperty(exports, "roleOrEnumDoor", { enumerable: true, get: function () { return roleEnumUtils_1.roleOrEnumDoor; } });
Object.defineProperty(exports, "roleOrEnumGate", { enumerable: true, get: function () { return roleEnumUtils_1.roleOrEnumGate; } });
exports.default = ChannelDetector_1.ChannelDetector;
var types_1 = require("./types");
Object.defineProperty(exports, "Types", { enumerable: true, get: function () { return types_1.Types; } });
Object.defineProperty(exports, "StateType", { enumerable: true, get: function () { return types_1.StateType; } });
