// @iobroker/type-detector/build/types.js
"use strict";
/**
 * Copyright 2018-2026 bluefox <dogafox@gmail.com>
 *
 * The MIT License (MIT)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StateType = exports.Types = void 0;
// Make sure when you add a new entry here to also add in lib/types with the
// end user facing name for documentation
/** These are the types of patterns as used externally and internally */
var Types;
(function (Types) {
    Types["unknown"] = "unknown";
    Types["airCondition"] = "airCondition";
    Types["blind"] = "blind";
    Types["blindButtons"] = "blindButtons";
    Types["button"] = "button";
    Types["buttonSensor"] = "buttonSensor";
    Types["camera"] = "camera";
    Types["chart"] = "chart";
    Types["cie"] = "cie";
    Types["ct"] = "ct";
    Types["dimmer"] = "dimmer";
    Types["door"] = "door";
    Types["fireAlarm"] = "fireAlarm";
    Types["floodAlarm"] = "floodAlarm";
    Types["gate"] = "gate";
    Types["hue"] = "hue";
    Types["humidity"] = "humidity";
    Types["illuminance"] = "illuminance";
    Types["image"] = "image";
    Types["info"] = "info";
    Types["instance"] = "instance";
    Types["light"] = "light";
    Types["location"] = "location";
    Types["locationOne"] = "locationOne";
    Types["lock"] = "lock";
    Types["media"] = "media";
    Types["motion"] = "motion";
    Types["percentage"] = "percentage";
    Types["rgb"] = "rgb";
    Types["rgbSingle"] = "rgbSingle";
    Types["rgbwSingle"] = "rgbwSingle";
    Types["slider"] = "slider";
    Types["socket"] = "socket";
    Types["temperature"] = "temperature";
    Types["thermostat"] = "thermostat";
    Types["vacuumCleaner"] = "vacuumCleaner";
    Types["volume"] = "volume";
    Types["volumeGroup"] = "volumeGroup";
    Types["warning"] = "warning";
    Types["weatherCurrent"] = "weatherCurrent";
    Types["weatherForecast"] = "weatherForecast";
    Types["window"] = "window";
    Types["windowTilt"] = "windowTilt";
})(Types || (exports.Types = Types = {}));
var StateType;
(function (StateType) {
    StateType["Number"] = "number";
    StateType["String"] = "string";
    StateType["Boolean"] = "boolean";
})(StateType || (exports.StateType = StateType = {}));
